src/RequestParser.cpp

```diff
@@ -0,0 +1,551 @@
+        std::string headerName;
+        std::string headerValue;
+        if (std::getline(headerStream, headerName, ':')) {
+			checkHeaderName(headerName);
+			// getline() removes trailing \r\n
+            std::getline(headerStream >> std::ws, headerValue);
+			headerValue = trimTrailingWhiteSpaces(headerValue);
+			checkContentLength(headerName, headerValue);
+            m_request.headers[headerName] = headerValue;
+        }
+    }
+	checkTransferEncoding();
+	if (headerLine != "\r") {
+		m_errorCode = 400;
+		throw std::runtime_error(ERR_MISS_CRLF);
+	}
+
+    // Step 3: Parse body (if any)
+	if (m_request.hasBody) {
+		std::string body;
+		while (std::getline(requestStream, body)) {
+			if (!m_request.body.empty())
+				body += "\n";
+			m_request.body += body;
+		}
+	}
+    return m_request;
+}
+
+/**
+ * @brief Parses the HTTP method from the request line.
+ *
+ * This function extracts the HTTP method from the beginning of the provided request line.
+ * It supports the "GET", "POST", and "DELETE" methods. The method is stored in `m_request.method`,
+ * and a corresponding bit flag is set in `m_requestMethod`. If the method is not recognized,
+ * an error code is set to 501 and a `std::runtime_error` is thrown.
+ *
+ * @param requestLine The request line string from which the method is to be parsed.
+ * @return A substring of `requestLine` starting from the character after the parsed method.
+ * @throws std::runtime_error If the method is not implemented, an error is thrown and the error code is set to 501.
+ */
+std::string	RequestParser::parseMethod(const std::string& requestLine)
+{
+	int	i = -1;
+	while (isalpha(requestLine[++i]))
+		m_request.method.push_back(requestLine[i]);
+
+	if (m_request.method == "GET")
+		m_requestMethod = 1 << 0;
+	else if (m_request.method == "POST")
+		m_requestMethod = 1 << 1;
+	else if (m_request.method == "DELETE")
+		m_requestMethod = 1 << 2;
+	else {
+		m_errorCode = 501;
+		throw std::runtime_error(ERR_METHOD_NOT_IMPLEMENTED);
+	}
+	return (requestLine.substr(i));
+}
+
+/**
+ * @brief Parses the URI from the request line.
+ *
+ * This function extracts the URI from the provided request line.
+ * It validates the URI, ensuring it starts with a slash ('/'),
+ * and checks each character for validity.
+ * If the URI contains a query ('?') or fragment ('#'), it delegates parsing to
+ * `parseUriQuery` and `parseUriFragment` respectively.
+ * If an invalid character is encountered, an error code is set, and
+ * a `std::runtime_error` is thrown.
+ *
+ * @param requestLine The request line string from which the URI is to be parsed.
+ * @return A substring of `requestLine` starting from the character after the parsed URI.
+ * @throws std::runtime_error If the URI is missing the initial slash,
+ *         contains invalid characters, or any other URI-related errors occur.
+ */
+std::string	RequestParser::parseUri(const std::string& requestLine)
+{
+	int	i = 0;
+	if (requestLine[i] != '/') {
```

Comment:
I wouldn't say that array access should be avoided completely when we're coding in C++. It depends on the use case, sometimes working with indices is more readable and less verbose than iterators.

In this case I find it more straightforward to transfer the current position inter-functional when passing it to the subfunctions to parse for the fragment or query optionally.

Reply:
Found the check: [clang-tidy](https://clang.llvm.org/extra/clang-tidy/checks/cppcoreguidelines/pro-bounds-constant-array-index.html) 
The problem seems to be less the [] access, but that "i" isn't constant - the compiler can't check 
Alternative would be to at least use .at() which is bounds checked.

I think this boils down to best practises. 

C and  C++  are considered unsafe because they for example allow not bounds checked access into memory. The core guidelines are written to warn from bad programming styles. [Here](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es42-keep-use-of-pointers-simple-and-straightforward) they list this classic loop as "BAD" in example three.

Reply:
Alright, in order to ensure no out-of-bounds lookups we can use an alternative which is protected. But what is this at()?
I know only of std::array::at which is a feature of C++11. In the example given it is used differently though:
```
void f2()
{
    int arr[COUNT];
    for (int i = 0; i < COUNT; ++i)
        at(arr, i) = i;
}
```
I haven't found anything resembling this syntax.

Reply:
I think the "most C++ style" would be 
```
std::string::const_iterator it = requestLine.begin();
if (*it != '/') {
  m_errorCode = 400;
  throw std::runtime_error(ERR_URI_MISS_SLASH);
}
```
Which shows intent that we don't want to modify something (const_iterator), and want to look inside the container. 
Now iterators are not per se bounds checked, but they convey intent. The clang-tidy warning essentially says: hey you use 'i', and 'i' can be anything.
I think to shut it up you have to use the string method at(), which would look like
```
int	i = 0;
if (requestLine.at(i) != '/') {
  m_errorCode = 400;
  throw std::runtime_error(ERR_URI_MISS_SLASH);
}
```

Reply:
Changed.